Feature/ring scanner

asdocoder opened the pull request:
Changes to a forked version of this library with the following changes:

- Added support for multiple BTSerialPortBindingServer objects to be instantiated at the same time (on different channels) by changing some static elements to non-static.
- Fixed up the handling of client disconnects so that we re-advertise to SDP to handle another connection when a client disconnects (or we disconnect the client).
- Added DisconnectClient and IsOpen methods available to the js object
- We only close the server socket when explicitly told to close.

Commit:
fix: PERCEPTION-2240 added support for multiple BTSerialPortBindingServer objects to be
instantiated at the same time (on different channels) by changing some static elements
to non-static.  Also fixed up the handling of client disconnects so that we re-advertise
to SDP to handle another connection when a client disconnects (or we disconnect the client).
We only close the server socket when explicitely told to close.
Added DisconnectClient and IsOpen methods available to the js object.

barabas6river committed:
fix: trigger build

Commit:
feat: fixed handling of scanner disconnect

Commit:
Merge branch 'feature/ring_scanner' of github.com:6RiverSystems/node-bluetooth-serial-port into feature/ring_scanner

barabas6river committed:
fix: trigger build

DrCodeWizzard commented on src/linux/BTSerialPortBindingServer.cc:
```diff
@@ -306,10 +288,13 @@ void BTSerialPortBindingServer::EIO_AfterRead(uv_work_t *req) {
-        if(baton->errorno == ECONNRESET || baton->size == 0){
+        if(baton->errorno == ECONNRESET || baton->errorno == ETIMEDOUT || baton->size == 0 || baton->isDisconnect){
+            baton->rfcomm->CloseClientSocket();
+            if (!baton->isClose) {
```
why would this happen?

asdocoder replied:
ETIMEDOUT is what I got when I either walked out of range with the scanner or did a battery pull.  I'm not sure when ECONNRESET gets triggered, but I left it in there since it was in the original code.

DrCodeWizzard commented on src/linux/BTSerialPortBindingServer.cc:
```diff
@@ -391,13 +372,11 @@ NAN_METHOD(BTSerialPortBindingServer::New) {
+    uv_mutex_init(&rfcomm->mWriteQueueMutex);
+    ngx_queue_init(&rfcomm->mWriteQueue);
+    listen_baton_t * baton = rfcomm->mListenBaton;
     rfcomm->Wrap(info.This());
```
where is info coming from?

asdocoder replied:
info is coming from the NAN_METHOD macro.  It's a parameter passed into all methods that are exposed to javascript.  It contains any arguments passed in from javascript, a pointer to this, and something to hold any return value(s).

barabas6river commented on src/linux/BTSerialPortBindingServer.cc:
```diff
@@ -345,30 +330,26 @@ void BTSerialPortBindingServer::Init(Handle<Object> target) {
 
     target->Set(Nan::New("BTSerialPortBindingServer").ToLocalChecked(), t->GetFunction());
-    target->Set(Nan::New("BTSerialPortBindingServer").ToLocalChecked(), t->GetFunction());
-    target->Set(Nan::New("BTSerialPortBindingServer").ToLocalChecked(), t->GetFunction());
```
Any idea why this line was duplicated in the original?

asdocoder replied:
No, the only thing I can think of was that there were 3 exposed methods before, and there's code right above it that had to be called once per method, so maybe this was a copy/paste error where the coder thought this needed to be per exposed method but never went back to try to change it (and then realized it wasn't specific per method)...

barabas6river commented:
Adding DanW as optional reviewer.

barabas6river reviewed (dismissed):
Looks good. Can you add the new DisconnectClient and IsOpen methods to the readme.md?

Commit:
documented changes in README.md

Commit:
Merge branch 'feature/ring_scanner' of github.com:6RiverSystems/node-bluetooth-serial-port into feature/ring_scanner

Commit:
updated doc in README.md

Commit:
feat: trigger build numbe change

asdocoder commented:
Thanks @barabas6river!  I've updated the README.md.  It already had isOpen (it was defined in the js, but only used local js objects - now it makes a call to the C++ object).  I added disconnectClient as well as the "disconnected" event (also fixed the formatting for the close events).  I thought I would be tricky and not add feat/fix to the checkin comment since the README.md isn't included in the build, but I see CircleCI still runs (and complains, because the version # didn't change).  Fixed now.
